### PR TITLE
Add MemoryExtensions.Count

### LINKED ISSUE #59466: [API Proposal]: MemoryExtensions.Count / number of occurrences of a value in a span
### Background and motivation

For a performance sensitive code I had to compute a string, for which I had to replace certain characters in a input string. So in order to use `string.Create` I had to know the final size upfront, for this I had to count the number of a specific char I wanted to replace (e.g replacing '[' with '<' & ' ' (2 chars), so the final size was old size + number of chars to replace found)

To my surprise there is currently no string.Count method that is vectorized or I've missed it... 
A simple for loop would've worked as well to count, but as seen [here](https://sergiopedri.medium.com/optimizing-string-count-all-the-way-from-linq-to-hardware-accelerated-vectorized-instructions-186816010ad9), it is slower than a vectorized count. (The link also shows an example implementation of a vectorized count and its improvements

### API Proposal

```C#
namespace System
{
    public sealed partial class MemoryExtensions
    {
        public static int Count<T>(this ReadOnlySpan<T> span, T value) where T : IEquatable<T>?;

        public static int Count<T>(this ReadOnlySpan<T> span, ReadOnlySpan<T> value) where T : IEquatable<T>?;

        public static int Count<T>(this Span<T> span, T value) where T : IEquatable<T>?;

        public static int Count<T>(this Span<T> span, ReadOnlySpan<T> value) where T : IEquatable<T>?;
    }
}
```     

### API Usage

```C#
// Instead of this
int count = 0;
int tmp = 0;
while ((tmp = charSpan.IndexOf('[', tmp)) >= 0)
{
    count++;
    tmp++;
}

// Or this
int count = 0;
for(int i =0; i < charSpan.Length; i++)
{
    if (charSpan[i] == '[')
    {
      count++;
    }
}


// We could do this in a more optimized and simplified version
int count = charSpan.Count('[');
```     

### Risks

None that I can think of.

### PR DESCRIPTION
**Pull request by bollhals**
Closes #59466 

Original implementation from @stephentoub.

Testcases added for single values and the same tests for ReadOnlySpans.
Types tested are byte, int, TInt and string.

Test cases are mostly copied / inspired by other similar methods like Contains.

### CODE REVIEW

**Comment by dnfadmin** [comment]
[CLA assistant check](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=80662) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=80662) before we can accept your contribution.<br/><br/>:x: bollhals [sign now](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=80662)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/dotnet/runtime?pullRequest=80662) it.</sub>

File: src/libraries/System.Memory/tests/ReadOnlySpan/Count.byte.cs [hunk 1]
```diff
@@ -0,0 +1,313 @@
...
+        
+        [Fact]
+        public static void DefaultFilledCount_Byte()
+        {
+            for (int length = 0; length <= byte.MaxValue; length++)
+            {
+                byte[] a = new byte[length];
+                ReadOnlySpan<byte> span = new ReadOnlySpan<byte>(a);
+
+                for (int i = 0; i < length; i++)
```
Comment (stephentoub) [inline src/libraries/System.Memory/tests/ReadOnlySpan/Count.byte.cs:37]: Why is this inner loop needed?

File: src/libraries/System.Memory/tests/ReadOnlySpan/Count.byte.cs [hunk 2]
```diff
@@ -0,0 +1,313 @@
...
+            ReadOnlySpan<byte> span = new ReadOnlySpan<byte>(Array.Empty<byte>());
+
+            int count = span.Count<byte>(new ReadOnlySpan<byte>(new byte[] { 0, 0 }));
+            Assert.Equal(0, count);
+        }
+        
+        [Fact]
+        public static void DefaultFilledCount_Byte()
+        {
+            for (int length = 0; length <= byte.MaxValue; length++)
```
Comment (stephentoub) [inline src/libraries/System.Memory/tests/ReadOnlySpan/Count.byte.cs:32]: We don't need to test all 256 byte values.  You could choose a lower max and/or just choose a few specific values to validate various border conditions.

File: src/libraries/System.Memory/tests/ReadOnlySpan/Count.byte.cs [hunk 3]
```diff
@@ -0,0 +1,313 @@
...
+        
+        [Fact]
+        public static void DefaultFilledCount_RosByte()
+        {
+            for (int length = 0; length <= byte.MaxValue; length++)
+            {
+                byte[] a = new byte[length];
+                ReadOnlySpan<byte> span = new ReadOnlySpan<byte>(a);
+
+                for (int i = 0; i < length; i++)
```
Comment (stephentoub) [inline src/libraries/System.Memory/tests/ReadOnlySpan/Count.byte.cs:55]: Same question; this inner loop doesn't look necessary... nothing in the body refers to i, so it's just running the same test repeatedly.

File: src/libraries/System.Memory/tests/ReadOnlySpan/Count.byte.cs [hunk 4]
```diff
@@ -0,0 +1,313 @@
...
+        [Fact]
+        public static void TestCount_Byte()
+        {
+            for (int length = 0; length <= byte.MaxValue; length++)
+            {
+                byte[] a = new byte[length];
+                for (int i = 0; i < length; i++)
+                {
+                    a[i] = (byte)(i + 1);
+                }
```
Comment (stephentoub) [inline src/libraries/System.Memory/tests/ReadOnlySpan/Count.byte.cs:74]:
```suggestion
                byte[] a = Enumerable.Range(1, length).Select(i => (byte)i).ToArray();
```

File: src/libraries/System.Memory/tests/ReadOnlySpan/Count.byte.cs [hunk 5]
```diff
@@ -0,0 +1,313 @@
...
+                ReadOnlySpan<byte> span = new ReadOnlySpan<byte>(a);
+
+                for (int targetIndex = 0; targetIndex < length; targetIndex++)
+                {
+                    byte target = a[targetIndex];
+
```
Comment (stephentoub) [inline src/libraries/System.Memory/tests/ReadOnlySpan/Count.byte.cs:80]:
```suggestion
                foreach (byte target in span)
                {
```

File: src/libraries/System.Memory/tests/ReadOnlySpan/Count.byte.cs [hunk 6]
```diff
@@ -0,0 +1,313 @@
...
+                }
+            }
+        }
+        
+        [Fact]
+        public static void TestCount_RosByte()
+        {
+            for (int length = 0; length <= byte.MaxValue; length++)
+            {
+                byte[] a = new byte[length];
```
Comment (stephentoub) [inline src/libraries/System.Memory/tests/ReadOnlySpan/Count.byte.cs:92]: Similar comments here and elsewhere

File: src/libraries/System.Memory/tests/ReadOnlySpan/Count.byte.cs [hunk 7]
```diff
@@ -0,0 +1,313 @@
...
+                for (int i = 0; i < length; i++)
+                {
+                    a[i] = (byte)(i + 1);
+                }
+                ReadOnlySpan<byte> span = new ReadOnlySpan<byte>(a);
+
+                for (int targetIndex = 0; targetIndex < length - 1; targetIndex++)
+                {
+                    ReadOnlySpan<byte> target = stackalloc byte[] { a[targetIndex], a[targetIndex + 1] };
```
Comment (stephentoub) [inline src/libraries/System.Memory/tests/ReadOnlySpan/Count.byte.cs:101]: stackalloc within a loop is dangerous and in a test like this we can just use new arrays...
```suggestion
                    ReadOnlySpan<byte> target = new byte[] { a[targetIndex], a[targetIndex + 1] };
```
Alternatively you could create the array once before the loop and then just repopulate it on each call

File: src/libraries/System.Memory/tests/ReadOnlySpan/Count.byte.cs [hunk 8]
```diff
@@ -0,0 +1,313 @@
...
+
+                    int count = span.Count(target);
+                    Assert.Equal(1, count);
```
Comment (stephentoub) [inline src/libraries/System.Memory/tests/ReadOnlySpan/Count.byte.cs:104]: In all of these cases where you're first storing the value and then asserting, you could instead just do e.g.
```suggestion
                    Assert.Equal(1, span.Count(target));
```

File: src/libraries/System.Memory/tests/ReadOnlySpan/Count.byte.cs [hunk 9]
```diff
@@ -0,0 +1,313 @@
...
+        }
+
+        [Fact]
+        public static void TestAlignmentCount_Byte()
+        {
+            byte[] array = new byte[4 * Vector<byte>.Count];
+            for (int i = 0; i < array.Length; i++)
+            {
+                array[i] = 5;
+            }
```
Comment (stephentoub) [inline src/libraries/System.Memory/tests/ReadOnlySpan/Count.byte.cs:195]:
```suggestion
            Array.Fill(array, 5);
```

File: src/libraries/System.Memory/tests/ReadOnlySpan/Count.byte.cs [hunk 10]
```diff
@@ -0,0 +1,313 @@
...
+                count = span.Count<byte>(5);
+                Assert.Equal(span.Length, count);
+            }
+        }
+        
+        [Fact]
+        public static void TestAlignmentCount_RosByte()
+        {
+            byte[] array = new byte[4 * Vector<byte>.Count];
+            for (int i = 0; i < array.Length; i++)
```
Comment (stephentoub) [inline src/libraries/System.Memory/tests/ReadOnlySpan/Count.byte.cs:214]: Ditto here and elsewhere

File: src/libraries/System.Memory/tests/ReadOnlySpan/Count.T.cs [hunk 11]
```diff
@@ -0,0 +1,402 @@
...
+            }
+        }
+
+        [Fact]
+        public static void MakeSureNoChecksForCountGoOutOfRange_TInt()
+        {
+            const int GuardValue = 77777;
+            const int GuardLength = 50;
+
+            void checkForOutOfRangeAccess(int x, int y)
```
Comment (stephentoub) [inline src/libraries/System.Memory/tests/ReadOnlySpan/Count.T.cs:184]: Nit: PascalCasing

File: src/libraries/System.Memory/tests/ReadOnlySpan/Count.T.cs [hunk 12]
```diff
@@ -0,0 +1,402 @@
...
+                Assert.Equal(2, count);
+            }
+        }
+
+        [Fact]
+        public static void TestMultipleMatchCount_RosString()
+        {
+            for (int length = 4; length < 32; length++)
+            {
+                string[] a = new string[length];
```
Comment (stephentoub) [inline src/libraries/System.Memory/tests/ReadOnlySpan/Count.T.cs:367]: It'd be good to have some string tests that validate:
1) It behaves like StringComparison.Ordinal, e.g. make sure things that would compare equivalent under a certain culture but not under ordinal don't compare equivalent, make sure things that would compare equivalent under OrdinalIgnoreCase but not under ordinal don't compare equivalent, etc.
2) That overlaps aren't counted, e.g. searching for "aa" in "aaaaa" should produce 2, not 4.

Comment (stephentoub) [reply src/libraries/System.Memory/tests/ReadOnlySpan/Count.T.cs:367]: Also, I didn't notice... are there any tests that exercise the path in the multiple values overload where it in turn just delegates to the single value overload?

File: src/libraries/System.Private.CoreLib/src/System/MemoryExtensions.cs [hunk 13]
```diff
@@ -3563,6 +3563,92 @@ private static (int StartInclusive, int EndExclusive) TrimSplitEntry(ReadOnlySpa
...
+        /// <returns>The number of times <paramref name="value"/> was found in the <paramref name="span"/>.</returns>
+        public static int Count<T>(this Span<T> span, ReadOnlySpan<T> value) where T : IEquatable<T>? =>
+            Count((ReadOnlySpan<T>)span, value);
+
+        /// <summary>Counts the number of times the specified <paramref name="value"/> occurs in the <paramref name="span"/>.</summary>
+        /// <typeparam name="T">The element type of the span.</typeparam>
+        /// <param name="span">The span to search.</param>
+        /// <param name="value">The value for which to search.</param>
+        /// <returns>The number of times <paramref name="value"/> was found in the <paramref name="span"/>.</returns>
+        public static int Count<T>(this ReadOnlySpan<T> span, ReadOnlySpan<T> value) where T : IEquatable<T>?
```
Comment (stephentoub) [inline src/libraries/System.Private.CoreLib/src/System/MemoryExtensions.cs:3633]: Do we have any tests for an empty value?  I think based on this implementation it would likely end up in an infinite loop.  We should ensure we have tests for it, and we probably need to special-case value.Length == 0 to just return 0 (or potentially throw an exception).

**Comment by stephentoub** [comment]
Thanks for working on this!

File: src/libraries/System.Private.CoreLib/src/System/SpanHelpers.T.cs [hunk 14]
```diff
@@ -3292,5 +3293,88 @@ internal static int LastIndexOfAnyExceptInRangeUnsignedNumber<T>(ref T searchSpa
...
+            ref T end = ref Unsafe.Add(ref current, length);
+            if (Vector128.IsHardwareAccelerated && length >= Vector128<T>.Count)
+            {
+                if (Vector256.IsHardwareAccelerated && length >= Vector256<T>.Count)
+                {
+                    Vector256<T> targetVector = Vector256.Create(value);
+                    ref T oneVectorAwayFromEndMinus1 = ref Unsafe.Subtract(ref end, Vector256<T>.Count - 1);
+                    do
+                    {
+                        count += BitOperations.PopCount(Vector256.Equals(Vector256.LoadUnsafe(ref current), targetVector).ExtractMostSignificantBits());
```
Comment (EgorBo) [inline src/libraries/System.Private.CoreLib/src/System/SpanHelpers.T.cs:3343]: A small note `ExtractMostSignificantBits` is quite expensive on ARM so it might be even cheaper to check Vector256.Equals being all zero first - at least that is what we do in all other places.

Although, we should make it faster with https://github.com/dotnet/runtime/issues/76047 so let's keep it as is

Comment (gfoidl) [inline src/libraries/System.Private.CoreLib/src/System/SpanHelpers.T.cs:3343]: Instead of doing a _popcount_ on each iteration you could use an accumulator register for the counts, then do the reduction after the loop.

(If not done in this PR, I (or anyone else) can do this in a separate PR.
I like bringing the functionality in, then optimize it when and where possible.)

File: src/libraries/System.Memory/tests/ReadOnlySpan/Count.byte.cs (hunk 1 above)
Comment (bollhals) [reply src/libraries/System.Memory/tests/ReadOnlySpan/Count.byte.cs:37]: I don't know, copied from DefaultFilledContains_Byte. Shall I remove them both?

Comment (stephentoub) [reply src/libraries/System.Memory/tests/ReadOnlySpan/Count.byte.cs]: Yup, that loop is also unnecessary.

File: src/libraries/System.Private.CoreLib/src/System/SpanHelpers.T.cs (hunk 14 above)
Comment (stephentoub) [reply src/libraries/System.Private.CoreLib/src/System/SpanHelpers.T.cs:3343]: Let's do it separately, with appropriate measurement.

File: src/libraries/System.Private.CoreLib/src/System/SpanHelpers.T.cs [hunk 15]
```diff
@@ -3292,5 +3293,88 @@ internal static int LastIndexOfAnyExceptInRangeUnsignedNumber<T>(ref T searchSpa
...
+                        current = ref Unsafe.Add(ref current, Vector256<T>.Count);
+                    }
+                    while (Unsafe.IsAddressLessThan(ref current, ref oneVectorAwayFromEndMinus1));
+
+                    if (Unsafe.IsAddressLessThan(ref current, ref Unsafe.Subtract(ref end, Vector128<T>.Count - 1)))
```
Comment (danmoseley) [inline src/libraries/System.Private.CoreLib/src/System/SpanHelpers.T.cs:3348]: @tannergooding given this pattern where the Vector256, Vector128, and presumably Vector512 implementations are otherwise identical -- do we have a pattern planned that would allow them to  be collapsed? this seems like a place where you want to just use Vector, but still get the benefits of a fixed size vector, without writing all this out.

**Comment by danmoseley** [comment]
Perhaps it was already mentioned above, but do you plan to follow up with a PR to add test cases to dotnet/performance?

File: src/libraries/System.Private.CoreLib/src/System/SpanHelpers.T.cs (hunk 15 above)
Comment (stephentoub) [reply src/libraries/System.Private.CoreLib/src/System/SpanHelpers.T.cs:3348]: We've talked about a few approaches:
- Allowing `Vector<T>` to play nicely with dynamic PGO
- Having an `IVector<T>` that would let you write the code once and call it three times with the three different vector sizes
- For specific kinds of situations, e.g. IndexOf, experiment with using a pattern based on an interface that lets you express just the kernel of the search for each of the vector sizes and have the calling API supply all the boilerplate around it.

Comment (tannergooding) [reply src/libraries/System.Private.CoreLib/src/System/SpanHelpers.T.cs:3348]: @danmoseley, yes. I need to take my findings from https://github.com/dotnet/runtime/pull/76423 and create an API proposal.

I have most of one complete, but not all of one yet

File: src/libraries/System.Private.CoreLib/src/System/MemoryExtensions.cs (hunk 13 above)
Comment (bollhals) [reply src/libraries/System.Private.CoreLib/src/System/MemoryExtensions.cs:3633]: There wasn't, and yes, it ended up in an infinite loop.

File: src/libraries/System.Memory/tests/ReadOnlySpan/Count.T.cs [hunk 16]
```diff
@@ -0,0 +1,400 @@
...
+
+            Assert.Equal(0, span.Count(0));
+        }
+        
+        [Fact]
+        public static void ZeroLengthCount_RosInt()
+        {
+            ReadOnlySpan<int> span = new ReadOnlySpan<int>(Array.Empty<int>());
+
+            Assert.Equal(0, span.Count<int>(new ReadOnlySpan<int>(new int[] { 0, 0 })));
```
Comment (stephentoub) [inline src/libraries/System.Memory/tests/ReadOnlySpan/Count.T.cs:23]:
```suggestion
            ReadOnlySpan<int> span = ReadOnlySpan<int>.Empty;
            for (int i = 0; i <= 2; i++)
            {
                Assert.Equal(0, span.Count(new ReadOnlySpan<int>(new int[i])));
            }
```

File: src/libraries/System.Memory/tests/ReadOnlySpan/Count.byte.cs [hunk 17]
```diff
@@ -0,0 +1,298 @@
...
+using Xunit;
+
+namespace System.SpanTests
+{
+    public static partial class ReadOnlySpanTests
+    {
+        [Fact]
+        public static void ZeroLengthCount_Byte()
+        {
+            ReadOnlySpan<byte> span = new ReadOnlySpan<byte>(Array.Empty<byte>());
```
Comment (stephentoub) [inline src/libraries/System.Memory/tests/ReadOnlySpan/Count.byte.cs:15]: Same comments as on the char tests

File: src/libraries/System.Memory/tests/TestHelpers.cs [hunk 18]
```diff
@@ -403,7 +403,7 @@ public static Memory<T> DangerousCreateMemory<T>(object obj, int offset, int len
         /// <summary>Creates a <see cref="ReadOnlyMemory{T}"/> with the specified values in its backing field.</summary>
         public static ReadOnlyMemory<T> DangerousCreateReadOnlyMemory<T>(object obj, int offset, int length) =>
             DangerousCreateMemory<T>(obj, offset, length);
-
+
```
Comment (stephentoub) [inline src/libraries/System.Memory/tests/TestHelpers.cs:406]:
```suggestion

```

**Review by stephentoub — APPROVED** [review]
Thanks!
